edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/planners/HealthRestartPlanner.cs

```diff
@@ -70,6 +73,49 @@ public async Task<Plan> CreateShutdownPlanAsync(ModuleSet current)
         {
             Events.LogDesired(desired);
             Events.LogCurrent(current);
+
+            List<ICommand> commands = new List<ICommand>();
+
+            // Create a grouping of desired and current modules based on their priority.
+            // We want to process all the modules in the deployment (desired modules) and also include the modules
+            // that are not specified in the deployment but are currently running on the device. This is so that
+            // their processing is done in the right priority order.
+            ILookup<uint, KeyValuePair<string, IModule>> desiredPriorityGroups = desired.Modules.ToLookup(x => x.Value.Priority);
+            ILookup<uint, KeyValuePair<string, IModule>> currentPriorityGroups = current.Modules.ToLookup(x => x.Value.Priority);
+            ImmutableSortedSet<uint> orderedPriorities = desiredPriorityGroups.Select(x => x.Key).Union(currentPriorityGroups.Select(x => x.Key)).ToImmutableSortedSet();
+            HashSet<string> processedDesiredMatchingCurrentModules = new HashSet<string>();
+
+            foreach (uint priority in orderedPriorities)
+            {
+                // The desired set is all the desired modules that have the priority of the current priority group being evaluated.
+                ModuleSet priorityBasedDesiredSet = ModuleSet.Create(desiredPriorityGroups[priority].Select(x => x.Value).ToArray());
+
+                // The current set is:
+                // - All the current modules that correspond to the desired modules present in the current priority group.
+                // - All the current modules that have the priority of the current priority group being evaluated which were not specified in the desired deployment config
+                //   -and- have not already been processed yet.
+                //   These are included so that they can be stopped and removed in the right priority order.
+                IEnumerable<KeyValuePair<string, IModule>> desiredMatchingCurrentModules = current.Modules.Where(x => priorityBasedDesiredSet.Modules.ContainsKey(x.Key));
```

Comment:
So, we need this in case the priority of a deployed module has changed?

Reply:
That's correct.